activerecord/lib/active_record/relation/finder_methods.rb

```diff
@@ -507,6 +508,7 @@ def find_some_ordered(ids)
         result = except(:limit, :offset).where(primary_key => ids).records
 
         if result.size == ids.size
+          raise ArgumentError, "Primary key is not included in the custom select clause" unless result.first.id
           result.in_order_of(:id, ids.map { |id| @klass.type_for_attribute(primary_key).cast(id) })
```

Comment:
That said, it's an orthogonal concern, but I think we should probably raise an error if `in_order_of` returns less elements than `result`.